src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/VideoLinkAppointmentLinker.kt

```diff
@@ -0,0 +1,118 @@
+  fun getBookingIdsOfUnlinkedAppointments(): List<Long> {
+    val bookingIds = videoLinkAppointmentRepository.bookingIdsOfUnlinkedAppointments()
+    log.info("Found unlinked VideoLinkAppointments for ${bookingIds.size} bookingIds")
+    return bookingIds
+  }
+
+  @Transactional
+  fun linkAppointmentsForOffenderBookingId(bookingId: Long) {
+    val bookings = videoLinkBookingsForOffenderBookingId(bookingId)
+
+    log.info("bookingId $bookingId: Creating ${bookings.size} VideoLinkBookings")
+
+    videoLinkBookingRepository.saveAll(bookings)
+  }
+
+  fun videoLinkBookingsForOffenderBookingId(bookingId: Long): List<VideoLinkBooking> {
+    val prisonAppointments = getPrisonVideoLinkAppointmentsForBookingId(bookingId)
+    val prisonAppointmentsById = prisonAppointments.associateBy { it.eventId }
+
+    val mainAppointments = videoLinkAppointmentRepository.unlinkedMainAppointmentsForBookingId(bookingId)
+
+    log.info("bookingId $bookingId: Found ${mainAppointments.size} MAIN VideoLinkAppointments")
+
+    val findPreAppointmentFor = preAppointmentFinder(bookingId, prisonAppointments)
+    val findPostAppointmentFor = postAppointmentFinder(bookingId, prisonAppointments)
+
+    return mainAppointments
+      .filter { prisonAppointmentsById.containsKey(it.appointmentId) }
+      .map { main ->
+        val mainPrisonAppointment = prisonAppointmentsById[main.appointmentId]
+
+        VideoLinkBooking(
+          main = main,
+          pre = findPreAppointmentFor(mainPrisonAppointment),
+          post = findPostAppointmentFor(mainPrisonAppointment),
+        )
+      }
+  }
+
+  fun getPrisonVideoLinkAppointmentsForBookingId(bookingId: Long): List<PrisonAppointment> {
+    val prisonAppointments = mutableListOf<PrisonAppointment>()
+    var offset = 0
+
+    do {
+      val chunk = prisonApiService
+        .getPrisonAppointmentsForBookingId(bookingId, offset, CHUNK_SIZE)
+        .filter { it.eventSubType == "VLB" }
+      prisonAppointments.addAll(chunk)
+      offset += chunk.size
+    } while (chunk.isNotEmpty())
+
+    log.info("bookingId $bookingId: Retrieved ${prisonAppointments.size} VLB appointments from prison-api")
+
+    return prisonAppointments
+  }
+
+  fun preAppointmentFinder(bookingId: Long, prisonAppointments: List<PrisonAppointment>): AppointmentFinder {
+    val appointmentsByAppointmentId = videoLinkAppointmentRepository
+      .unlinkedPreAppointmentsForBookingId(bookingId)
+      .associateBy { it.appointmentId }
+
+    val appointmentsByEndTime = prisonAppointments
+      .filter { appointmentsByAppointmentId.containsKey(it.eventId) }
+      .associateBy({ it.endTime }, { appointmentsByAppointmentId[it.eventId] })
+      .toMutableMap()
+
+    log.info("bookingId $bookingId: Found ${appointmentsByAppointmentId.size} PRE VideoLinkAppointments, Matched ${appointmentsByEndTime.size} with prison appointments")
+
+    return ({ mainAppointment -> appointmentsByEndTime.remove(mainAppointment?.startTime) })
+  }
```

Comment:
Don't want a pre/post appointment to be associated with more than one main appointment so remove when matched.